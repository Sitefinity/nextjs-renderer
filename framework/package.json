{
  "name": "sitefinity-react-renderer-framework",
  "version": "0.1.0",
  "private": true,
  "dependencies": {
    "@types/react": "18.2.28",
    "react": "18.2.0",
    "typescript": "5.2.2"
  },
  "scripts": {
    "build": "tsc"
  }
}
