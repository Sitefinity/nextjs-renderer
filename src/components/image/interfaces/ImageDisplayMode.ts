export enum ImageDisplayMode {
    Responsive = 'Responsive',
    OriginalSize = 'OriginalSize',
    Thumbnail = 'Thumbnail',
    CustomSize = 'CustomSize',
}
