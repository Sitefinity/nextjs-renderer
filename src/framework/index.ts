export * from './editor';
export * from './rest-sdk';
