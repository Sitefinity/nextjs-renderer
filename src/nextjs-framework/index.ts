export * from './pages/render-page';
export * from './pages/render-widget';
export * from './pages/utils';
