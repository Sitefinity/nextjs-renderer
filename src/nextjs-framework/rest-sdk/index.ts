export * from './rest-service';
export * from './root-url.service';
export * from './service-metadata';
export * from './dto';
export * from './filters';
export * from './services';
