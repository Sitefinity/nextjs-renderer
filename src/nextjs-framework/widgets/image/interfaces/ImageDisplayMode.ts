import { ChoiceItem } from '@progress/sitefinity-widget-designers';

export enum ImageDisplayMode {
    Responsive = 'Responsive',
    OriginalSize = 'OriginalSize',
    Thumbnail = 'Thumbnail',
    CustomSize = 'CustomSize',
}

export const ImageDisplayChoices: ChoiceItem[] = [
    { Value: 'Responsive'},
    { Value: 'OriginalSize', Title: 'Original size'},
    { Value: 'Thumbnail', Title: 'Use thumbnail'},
    { Value: 'CustomSize', Title: 'Custom size'}
];
