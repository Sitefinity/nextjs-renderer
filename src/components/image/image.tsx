import React from 'react';
import { StyleGenerator } from '../styling/style-generator.service';
import { OffsetStyle } from '../styling/offset-style';
import { generateAnchorAttrsFromLink, htmlAttributes } from 'sitefinity-react-framework/widgets/attributes';
import { classNames } from 'sitefinity-react-framework/utils/classNames';
import { WidgetContext } from 'sitefinity-react-framework/widgets/widget-context';
import { LinkModel } from 'sitefinity-react-framework/interfaces/LinkModel';
import { RestService, RestSdkTypes } from 'sitefinity-react-framework/sdk/rest-service';
import { ImageTag } from './image-tag';
import { ImageClickAction } from './interfaces/ImageClickAction';
import { ImageDisplayMode } from './interfaces/ImageDisplayMode';

const imageWrapperClass = 'd-inline-block';

export async function Image(props: WidgetContext<ImageEntity>) {
    const entity = {
        ImageSize: ImageDisplayMode.Responsive,
        ...props.model.Properties
    };
    const dataAttributes = htmlAttributes(props);
    const defaultClass = classNames(imageWrapperClass, entity.CssClass);
    const marginClass = entity.Margins && StyleGenerator.getMarginClasses(entity.Margins);
    const anchorAttributes = generateAnchorAttrsFromLink(entity.ActionLink);
    dataAttributes['className'] = classNames(
        defaultClass,
        marginClass
    );

    if(props.requestContext && props.requestContext.isEdit) {
        dataAttributes['data-sfemptyicon'] = 'picture-o';
        dataAttributes['data-sfemptyiconaction'] = 'Edit';
    }

    dataAttributes['data-sfemptyicontext'] = 'Select image';
    dataAttributes['data-sfhasquickeditoperation'] = true;
    let imageItem = null;
    if (entity.Item && entity.Item.Id){
        imageItem =  await RestService.getItemWithFallback(RestSdkTypes.Image, entity.Item.Id.toString(), entity.Item.Provider);
    }

    if (!imageItem) {
        return null;
    }

    const isSvg = imageItem.MimeType === 'image/svg+xml';
    const hasZeroDimensions = imageItem.Width === 0 && imageItem.Height === 0;
    let width = isSvg && hasZeroDimensions ? null : imageItem.Width;
    let height = isSvg && hasZeroDimensions ? null : imageItem.Height;
    let thumbnails;
    let selectedImageUrl =  imageItem.Url;

    if (imageItem.Thumbnails) {
        thumbnails = imageItem.Thumbnails.sort((a: any,b: any) => a.Width - b.Width);
        if (entity.ImageSize === ImageDisplayMode.Thumbnail && imageItem.Thumnail) {
            let selectedThumbnail = thumbnails.find((t: any) => t.Title === entity.Thumnail.Name);
            selectedImageUrl = imageItem.Thumnail.Url;

            if (selectedThumbnail) {
                selectedImageUrl = selectedThumbnail.Url;
                width = selectedThumbnail.Width;
                height = selectedThumbnail.Height;
            }
        }
    }

    // Thumbnails for images imported from DAM providers are not stored in Sitefinity and we do not know their width and height.
    // We have to set width and heigth to null otherwise the image is zoomed and selected thumbnail not applied correctly.
    if (entity.ImageSize === ImageDisplayMode.Thumbnail && imageItem.IsDamMedia) {
        width = null;
        height = null;
    }

    if (entity.CustomSize && entity.ImageSize === ImageDisplayMode.CustomSize) {
        width = entity.CustomSize.Width;
        height = entity.CustomSize.Height;
    }

    const imageModel = {
        Title: imageItem.Title,
        AlternativeText: imageItem.AlternativeText,
        ...entity,
        Width: width,
        Height: height,
        Thumbnails: thumbnails,
        SelectedImageUrl: selectedImageUrl
    };

    return  entity.ClickAction === ImageClickAction.OpenOriginalSize
                ? <a href={entity.Item.Url}
                    {...dataAttributes}>
                  {<ImageTag imageModel={imageModel} />}
                </a>
                : entity.ClickAction === ImageClickAction.OpenLink && entity.ActionLink!.href
                    ?  <a {...anchorAttributes} {...dataAttributes}>
                      {<ImageTag imageModel={imageModel}  />}
                    </a>
                    :  <ImageTag imageModel={imageModel} className={entity.CssClass}
                        {...dataAttributes} />
             ;
}

export interface ImageEntity {
    Item?: any;
    Attributes?: any[];
    Margins?: OffsetStyle;
    Title?: string;
    AlternativeText?: string;
    CssClass?: string;
    ClickAction?: ImageClickAction;
    ActionLink?: LinkModel;
    ImageSize?: ImageDisplayMode;
    FitToContainer: boolean;
    CustomSize?: any;
    Thumnail?: any;
    ViewName?: string;
}
