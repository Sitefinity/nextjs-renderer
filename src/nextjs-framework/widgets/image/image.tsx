import React from 'react';
import { StyleGenerator } from '../styling/style-generator.service';
import { OffsetStyle } from '../styling/offset-style';
import { ImageTag } from './image-tag';
import { ImageClickAction } from './interfaces/ImageClickAction';
import { ImageDisplayMode } from './interfaces/ImageDisplayMode';
import { WidgetContext, htmlAttributes, classNames, generateAnchorAttrsFromLink, LinkModel } from '../../editor';
import { RestService, RestSdkTypes, ThumbnailItem, SdkItem, ImageItem } from '../../rest-sdk';

const imageWrapperClass = 'd-inline-block';

export async function Image(props: WidgetContext<ImageEntity>) {
    const entity = {
        ImageSize: ImageDisplayMode.Responsive,
        ...props.model.Properties
    };
    const dataAttributes = htmlAttributes(props);
    const defaultClass = classNames(imageWrapperClass, entity.CssClass);
    const marginClass = entity.Margins && StyleGenerator.getMarginClasses(entity.Margins);
    const anchorAttributes = generateAnchorAttrsFromLink(entity.ActionLink);
    dataAttributes['className'] = classNames(
        defaultClass,
        marginClass
    );

    dataAttributes['data-sfhasquickeditoperation'] = true;
    let imageItem = null;
    if (entity.Item && entity.Item.Id) {
        imageItem =  await RestService.getItemWithFallback(RestSdkTypes.Image, entity.Item.Id.toString(), entity.Item.Provider);
    }

    if (!imageItem) {
        return (<div {...dataAttributes} />);
    }

    const isSvg = imageItem.MimeType === 'image/svg+xml';
    const hasZeroDimensions = imageItem.Width === 0 && imageItem.Height === 0;
    let width = isSvg && hasZeroDimensions ? null : imageItem.Width;
    let height = isSvg && hasZeroDimensions ? null : imageItem.Height;
    let thumbnails;
    let selectedImageUrl =  imageItem.Url;

    if (imageItem.Thumbnails) {
        thumbnails = imageItem.Thumbnails.sort((a: ThumbnailItem,b: ThumbnailItem) => a.Width - b.Width);
        if (entity.ImageSize === ImageDisplayMode.Thumbnail && imageItem.Thumnail) {
            let selectedThumbnail = thumbnails.find((t: ThumbnailItem) => t.Title === entity.Thumnail!.Name);
            selectedImageUrl = imageItem.Thumnail.Url;

            if (selectedThumbnail) {
                selectedImageUrl = selectedThumbnail.Url;
                width = selectedThumbnail.Width;
                height = selectedThumbnail.Height;
            }
        }
    }

    // Thumbnails for images imported from DAM providers are not stored in Sitefinity and we do not know their width and height.
    // We have to set width and heigth to null otherwise the image is zoomed and selected thumbnail not applied correctly.
    if (entity.ImageSize === ImageDisplayMode.Thumbnail && imageItem.IsDamMedia) {
        width = null;
        height = null;
    }

    if (entity.CustomSize && entity.ImageSize === ImageDisplayMode.CustomSize) {
        width = entity.CustomSize.Width;
        height = entity.CustomSize.Height;
    }

    const imageModel = {
        Title: imageItem.Title,
        AlternativeText: imageItem.AlternativeText,
        ...entity,
        Width: width,
        Height: height,
        Thumbnails: thumbnails,
        SelectedImageUrl: selectedImageUrl
    };

    return  entity.ClickAction === ImageClickAction.OpenOriginalSize
                ? <a href={entity.Item!.Url}
                    {...dataAttributes}>
                  {<ImageTag imageModel={imageModel} />}
                </a>
                : entity.ClickAction === ImageClickAction.OpenLink && entity.ActionLink!.href
                    ?  <a {...anchorAttributes} {...dataAttributes}>
                      {<ImageTag imageModel={imageModel}  />}
                    </a>
                    :  <ImageTag imageModel={imageModel} className={entity.CssClass}
                        {...dataAttributes} />
             ;
}

export interface ImageEntity {
    Item?: ImageItem;
    Attributes?: { [key: string]: Array<{ Key: string, Value: string}> };
    Margins?: OffsetStyle;
    Title?: string;
    AlternativeText?: string;
    CssClass?: string;
    ClickAction?: ImageClickAction;
    ActionLink?: LinkModel;
    ImageSize?: ImageDisplayMode;
    FitToContainer: boolean;
    CustomSize?: { Width: number, Height: number};
    Thumnail?: ThumbnailItem;
    ViewName?: string;
}
