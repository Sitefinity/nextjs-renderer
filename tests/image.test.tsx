import { waitFor } from '@testing-library/react';
import { WidgetTester } from './framework/widget-tester';
import { RestSdkTypes, RestService, SdkItem } from '../src/nextjs-framework/rest-sdk';
import fs from 'fs';
import { ImageEntity } from '../src/nextjs-framework/widgets/image/image';
import { ImageDisplayMode } from '../src/nextjs-framework/widgets/image/interfaces/ImageDisplayMode';

let sdkItem: SdkItem;
beforeAll(async () => {
    const base64Image = fs.readFileSync('./tests/data/1.jpg', { encoding: 'base64' });

    let libraryId = '4BA7AD46-F29B-4e65-BE17-9BF7CE5BA1FB';
    sdkItem = await RestService.uploadItem({
        Title: 'Image',
        Type: RestSdkTypes.Image,
        ContentType: 'image/jpeg',
        FileName: 'test.jpg',
        ParentId: libraryId,
        BinaryData: base64Image
    });
});

test('Image rendered with original size', async () => {
    await WidgetTester.testWidgetRender<ImageEntity>({
        name: 'SitefinityImage',
        properties: {
            Item: {
                Id: sdkItem.Id,
                Provider: sdkItem.Provider
            },
            FitToContainer: false,
            ImageSize: ImageDisplayMode.OriginalSize,
            Title: 'custom title'
        },
        assert: async (element) => {
            await waitFor(() => {
                expect(element).toMatchSnapshot();
            });
        }
    });
});

test('Image with selected image item', async () => {
    await WidgetTester.testWidgetRender({
        name: 'SitefinityImage',
        properties: {
            Item: {
                Id: sdkItem.Id,
                Provider: sdkItem.Provider
            }
        },
        assert: async (element) => {
            await waitFor(() => {
                const sourceElement = element.querySelector('source');

                // remove dynamic elements from snapshot assert
                expect(sourceElement?.getAttribute('srcset')).toBeDefined();
                sourceElement?.removeAttribute('srcset');

                const imgElement = element.querySelector('img');
                expect(imgElement?.getAttribute('src')).toBeDefined();
                imgElement?.removeAttribute('src');

                expect(element).toMatchSnapshot();
            });
        }
    });
});
