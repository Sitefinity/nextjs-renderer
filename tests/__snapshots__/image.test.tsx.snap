// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image rendered with original size 1`] = `
<div>
  <span>
    <img
      loading="lazy"
      src="https://systemvalforsnapshot"
      title="custom title"
    />
  </span>
</div>
`;

exports[`Image with selected image item 1`] = `
<div>
  <picture
    class="d-inline-block"
  >
    <source
      height="160"
      media="(max-width: 160px)"
      srcset="https://systemvalforsnapshot"
      type="image/jpeg"
      width="160"
    />
    <img
      alt=""
      class="d-inline-block"
      loading="lazy"
      src="https://systemvalforsnapshot"
      title="Image"
    />
  </picture>
</div>
`;

exports[`image css class rendered (Image_CssClass_Rendered) 1`] = `
<div>
  <span>
    <img
      class="myclass"
      loading="lazy"
      src="https://systemvalforsnapshot"
      title="custom title"
    />
  </span>
</div>
`;

exports[`image rendered with original size 1`] = `
<div>
  <span>
    <img
      loading="lazy"
      src="https://systemvalforsnapshot"
      title="custom title"
    />
  </span>
</div>
`;

exports[`image with selected image item(Image_OriginalSize_Rendered) 1`] = `
<div>
  <picture
    class="d-inline-block undefined"
  >
    <source
      height="160"
      media="(max-width: 160px)"
      srcset="https://systemvalforsnapshot"
      type="image/jpeg"
      width="160"
    />
    <img
      class="img-fluid"
      loading="lazy"
      src="https://systemvalforsnapshot"
      title="Image"
    />
  </picture>
</div>
`;
